lib/util.js

```diff
@@ -222,6 +222,44 @@ function urlToPath(path) {
     };
 }
 
+/**
+ * URL encodes each part of a URL path individually in order to support special
+ * characters # and &. Only http/https URLs are supported. Ensures that
+ * encoding happens only once, first by decoding and then re-encoding path parts.
+ *
+ * @param {String} path absolute URI whose path will be URL encoded. Supplied path
+ *  must be an absolute URL with protocal and host. For example:
+ *  https://host/path/to-be/encoded/myAsset.jpg
+ * @returns {String} absolute URI whose path has been URL encoded.
+ */
+function encodeUrlPath(path, base) {
+
+    const url = new URL(path, base);
+
+    if (url.protocol !== 'http:' && url.protocol !== 'https:' ) {
+        throw new UnsupportedFileUploadError(`only http/https urls are supported; url: ${path}`);
+    }
+
+    // retrieve URL path by taking section after origin - supports http paths with the # special
+    //  character, which otherwise are split between URL.pathname and URL.hash; special precautions
+    //  taken to remove ? query strings
+    let urlPath;
+    const queryStringIndex = url.toString().indexOf('?');
+    if (queryStringIndex === -1) {
+        urlPath = url.toString().substring(url.origin.length);
+    } else {
+        urlPath = url.toString().substring(url.origin.length, queryStringIndex);
+    }
+
+    // ensures URL is encoded only once by decoding and re-encoding
+    const encodedUrlArray = urlPath.split('/').map(part => {
+        part = decodeURIComponent(part);
+        return encodeURIComponent(part);
+    });
+
+    return `${url.origin}${encodedUrlArray.join('/')}`;
+}
```

Comment:
I'm confused by this function. URLs are supposed to be properly encoded, but this function suggests they are not?

If they are not, it should be fixed where the URLs are created and not later on because we will end up in complex situations where URLs end up being double-encoded which are impossible to do correctly.

The only reason we have the `FailUnsupportedAssets` function in here is to address AEM specific rules around which characters can be used in filenames. It's not to compensate for improperly encoded URLs.

Reply:
There are 2 URLs that need to be encoded during the AEM direct binary upload:

1. What becomes the `initiateUpload` URL, which a client (e.g. Asset Link) can properly encode without any required changes.
2. The `completeURI`, which is returned from AEM in an un-encoded form (see the sample response below). I agree that it would be cleanest to encode this URL coming out of AEM, however in speaking with @mfrisbey we believed it would be fastest to address in this library. 

What do you think? Is this the correct place to be encoding these URLs?

This function does decode, and then re-encode path parts to prevent double encoding. I agree that it is a risk. There is [one unit test](https://github.com/adobe/node-httptransfer/pull/130/files#diff-1d615d2a41f634be0b2c2c29d725eb0819df03577ec9cc527e561798f034ebe1R165) demonstrating this.

---

For reference, sample `initiateUpload` request:

`POST` to `https://author-p-e-cmstg.adobeaemcloud.com/content/dam/tcallsen-test/%23%26%2B()%20spec%20char%20path.initiateUpload.json` (AEM instance with special characters FT enabled)

Response:
```
{
	"folderPath": "/content/dam/tcallsen-test/#&+() spec char path",
	"files": [{
		"fileName": "# Cambridge1.jpg",
		"uploadURIs": ["https://author-p-e-cmstg.adobeaemcloud.com/abbreviated-block"],
		"uploadToken": "token",
		"mimeType": "image/jpeg",
		"minPartSize": 10485760,
		"maxPartSize": 104857600
	}],
	"completeURI": "/content/dam/tcallsen-test/#&+() spec char path.completeUpload.json"
}
```